core/src/main/java/com/linecorp/armeria/server/DefaultPathMappingContext.java

```diff
@@ -0,0 +1,246 @@
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+import javax.annotation.Nullable;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Splitter;
+import com.google.common.collect.ImmutableList;
+
+import com.linecorp.armeria.common.HttpHeaderNames;
+import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.common.HttpMethod;
+import com.linecorp.armeria.common.MediaType;
+import com.linecorp.armeria.common.MediaTypeSet;
+
+/**
+ * Holds the parameters which are required to find a service available to handle the request.
+ */
+final class DefaultPathMappingContext implements PathMappingContext {
+    private static final Logger logger = LoggerFactory.getLogger(DefaultPathMappingContext.class);
+
+    static final List<MediaType> ANY_TYPE = ImmutableList.of(MediaType.ANY_TYPE);
+
+    private static final Splitter ACCEPT_SPLITTER = Splitter.on(',').trimResults();
+
+    private final VirtualHost virtualHost;
+    private final String hostname;
+    private final HttpMethod method;
+    private final String path;
+    private final String query;
+    private final MediaType consumeType;
+    private final List<MediaType> produceTypes;
+    private final List<Object> summary;
+    private Throwable delayedCause;
+
+    DefaultPathMappingContext(VirtualHost virtualHost, String hostname,
+                              HttpMethod method, String path, @Nullable String query,
+                              @Nullable MediaType consumeType, @Nullable List<MediaType> produceTypes) {
+        this.virtualHost = requireNonNull(virtualHost, "virtualHost");
+        this.hostname = requireNonNull(hostname, "hostname");
+        this.method = requireNonNull(method, "method");
+        this.path = requireNonNull(path, "path");
+        this.query = query;
+        this.consumeType = consumeType;
+        this.produceTypes = produceTypes;
+        summary = generateSummary(this);
+    }
+
+    @Override
+    public VirtualHost virtualHost() {
+        return virtualHost;
+    }
+
+    @Override
+    public String hostname() {
+        return hostname;
+    }
+
+    @Override
+    public HttpMethod method() {
+        return method;
+    }
+
+    @Override
+    public String path() {
+        return path;
+    }
+
+    @Nullable
+    @Override
+    public String query() {
+        return query;
+    }
+
+    @Nullable
+    @Override
+    public MediaType consumeType() {
+        return consumeType;
+    }
+
+    @Nullable
+    @Override
+    public List<MediaType> produceTypes() {
+        return produceTypes;
+    }
+
+    @Override
+    public List<Object> summary() {
+        return summary;
+    }
+
+    @Override
+    public void delayThrowable(Throwable delayedCause) {
+        // Update with the last cause
+        this.delayedCause = delayedCause;
```

Comment:
> Is it OK to replace the old cause?

Yes, the priority of the cause should be controlled by the caller for now. Currently it is only called by `HttpHeaderPathMapping`, and the caller has its own precedence of the cause: `406 Not Acceptable` or `415 Unsupported Media Type` > `405 Method Not Allowed`.
If this method is more widely used than now, we need to discuss the priority of the cause.